awx_collection/plugins/module_utils/tower_module.py

```diff
@@ -0,0 +1,240 @@
+                    if type(config_data) is not dict:
+                        raise AssertionError("The yaml config file is not properly formatted as a dict.")
+                    try_config_parsing = False
+
+                except(AttributeError, yaml.YAMLError, AssertionError):
+                    try_config_parsing = True
+
+            if try_config_parsing:
+                # TowerCLI used to support a config file with a missing [general] section by prepending it if missing
+                if '[general]' not in config_string:
+                    config_string = '[general]\n{0}'.format(config_string)
+
+                config = ConfigParser()
+
+                try:
+                    placeholder_file = StringIO(config_string)
+                    # py2 ConfigParser has readfp, that has been deprecated in favor of read_file in py3
+                    # This "if" removes the deprecation warning
+                    if hasattr(config, 'read_file'):
+                        config.read_file(placeholder_file)
+                    else:
+                        config.readfp(placeholder_file)
+
+                    # If we made it here then we have values from reading the ini file, so let's pull them out into a dict
+                    config_data = {}
+                    for honorred_setting in self.short_params:
+                        try:
+                            config_data[honorred_setting] = config.get('general', honorred_setting)
+                        except NoOptionError:
+                            pass
+
+                except Exception as e:
+                    raise ConfigFileException("An unknown exception occured trying to ini load config file: {0}".format(e))
+
+        except Exception as e:
+            raise ConfigFileException("An unknown exception occured trying to load config file: {0}".format(e))
+
+        # If we made it here, we have a dict which has values in it from our config, any final settings logic can be performed here
+        for honorred_setting in self.short_params:
+            if honorred_setting in config_data:
+                # Veriffy SSL must be a boolean
+                if honorred_setting == 'verify_ssl':
+                    if type(config_data[honorred_setting]) is str:
+                        setattr(self, honorred_setting, strtobool(config_data[honorred_setting]))
+                    else:
+                        setattr(self, honorred_setting, bool(config_data[honorred_setting]))
+                else:
+                    setattr(self, honorred_setting, config_data[honorred_setting])
+
+
+    def logout(self):
+        # This method is intended to be overridden
+        pass
+
+    def fail_json(self, **kwargs):
+        # Try to log out if we are authenticated
+        self.logout()
+        if self.error_callback:
+            self.error_callback(**kwargs)
```

Comment:
This is another hack which I thought separating the classes would help solve. The `TowerAPIModule` class could just pass `self.fail_json` as the error callback, instead of having `self.fail_json` be the default behavior, which isn't valid for non-module plugins.